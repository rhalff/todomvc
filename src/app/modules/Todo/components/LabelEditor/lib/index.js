// @flow
export { default as cleanText } from './cleanText'
