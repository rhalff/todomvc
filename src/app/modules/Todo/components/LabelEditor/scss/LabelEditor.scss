@import 'vars';

.labelEditor {
  div {
    white-space: pre-line;
    word-break: break-all;
    padding: 15px 60px 15px 15px;
    line-height: 1.2;
    transition: color 0.4s;
    /* user-select: none; */

    &.completed {
      color: $labelEditorColor;
      text-decoration: line-through;
    }
  }

  &:hover > *[contenteditable="true"] {
    background-color: $labelEditorBackgroundColor;
  }
}
