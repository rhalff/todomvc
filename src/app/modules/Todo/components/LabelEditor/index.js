// @flow
import React, { Component } from 'react'
import styles from './scss/LabelEditor.scss'
import { cleanText } from './lib'

type Props = {
  onSave: Function,
  text: string
}

type State = {
  editMode: boolean,
  text?: string
}

class LabelEditor extends Component<Props, State> {
  labelEditor: ?HTMLElement
  state = {
    editMode: false,
    text: ''
  }

  onDoubleClick = () => {
    if (!this.state.editMode) {
      this.setState(
        () => ({
          ...this.state,
          editMode: true
        }),
        () => {
          if (this.labelEditor) {
            this.labelEditor.focus()
          }
        }
      )
    }
  }

  componentWillReceiveProps(nextProps: Props) {
    this.setState({
      text: nextProps.text,
      editMode: false
    })
  }

  onKeyDown = (event: SyntheticEvent<HTMLElement>) => {
    const { onSave } = this.props

    if (this.labelEditor) {
      const text = cleanText(String(this.labelEditor.innerText))

      if (event.key === 'Enter') {
        const newState: State = {
          editMode: false
        }

        if (text) {
          onSave(text)
          newState.text = text
        } else {
          this.revertDomManipulation()
        }

        this.setState(newState)
      }
    }
  }

  revertDomManipulation() {
    if (this.labelEditor) {
      this.labelEditor.innerHTML = this.props.text
    }
  }

  render() {
    const { editMode, text } = this.state

    return (
      <div className={styles.labelEditor}>
        <div
          ref={el => (this.labelEditor = el)}
          contentEditable={editMode}
          onKeyDown={this.onKeyDown}
          onDoubleClick={this.onDoubleClick}
        >
          {text || this.props.text}
        </div>
      </div>
    )
  }
}

export default LabelEditor
