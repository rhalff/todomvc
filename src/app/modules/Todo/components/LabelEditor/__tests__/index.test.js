// @flow
import React from 'react'
import LabelEditor from '../'
import renderer from 'react-test-renderer'

test('Can show CheckBox', () => {
  const component = renderer.create(
    <LabelEditor onSave={() => {}} text="Test LabelEditor" />
  )

  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('Can show CheckBox (checked)', () => {
  const component = renderer.create(<LabelEditor onChange={() => {}} checked />)

  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
