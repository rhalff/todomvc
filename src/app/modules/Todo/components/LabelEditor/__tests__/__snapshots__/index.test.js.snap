// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Can show CheckBox (checked) 1`] = `
<div
  className="labelEditor"
>
  <div
    contentEditable={false}
    onDoubleClick={[Function]}
    onKeyDown={[Function]}
  />
</div>
`;

exports[`Can show CheckBox 1`] = `
<div
  className="labelEditor"
>
  <div
    contentEditable={false}
    onDoubleClick={[Function]}
    onKeyDown={[Function]}
  >
    Test LabelEditor
  </div>
</div>
`;
